src/main/webapp/app/overview/submission-result-status.component.ts

```diff
@@ -46,9 +47,20 @@ export class SubmissionResultStatusComponent implements OnChanges {
             const quizExercise = this.exercise as QuizExercise;
             this.uninitialized = ArtemisQuizService.isUninitialized(quizExercise);
             this.quizNotStarted = ArtemisQuizService.notStarted(quizExercise);
+            this.shouldShowResult = !!this.studentParticipation?.results?.length;
         } else {
             this.uninitialized = !afterDueDate && !this.studentParticipation;
-            this.notSubmitted = !!this.studentParticipation && !this.studentParticipation.submissions?.length;
+            this.notSubmitted = afterDueDate && !!this.studentParticipation && !this.studentParticipation.submissions?.length;
+
+            if (this.exercise.type === ExerciseType.PROGRAMMING) {
+                const initializationStatesForResult = [InitializationState.INITIALIZED, InitializationState.INACTIVE, InitializationState.FINISHED];
+                this.shouldShowResult =
+                    (!!this.studentParticipation?.results?.length || !afterDueDate) &&
+                    !!this.studentParticipation?.initializationState &&
+                    initializationStatesForResult.includes(this.studentParticipation.initializationState);
+            } else {
+                this.shouldShowResult = this.studentParticipation?.initializationState === InitializationState.FINISHED;
+            }
```

Comment:
```suggestion
            this.setShouldShowResult();
```
I think this block is big enough to be worth moving it into a separate method.